Fix Timestamp rounding

alimcmaster1 opened the pull request:
- Closes #21262
- Tests added (thanks  @Safrone [PR](https://github.com/pandas-dev/pandas/pull/21265))

This change-set is to avoid rounding a timestamp when the timestamp is a multiple of the frequency string passed in.

"Values" param passed into round_ns can either be a np array or int. So relevant handling added for both.

FYI I havn't used Cython much before so keen to get peoples thoughts/feedback.

Thanks

stephenpascoe committed:
ENH:  Function to walk the group hierarchy of a PyTables HDF5 file.

closes #10143

alimcmaster1 committed:
DOC: MultiIndex Fixes (#21414)

jorisvandenbossche committed:
DOC: fix grammar of deprecation message (#21421)

bashtage committed:
MAINT: Deprecate encoding from stata reader/writer (#21400)

Deprecate the encoding parameter from all Stata reading and writing
methods and classes.  The encoding depends only on the file format and
cannot be changed by users.

gfyoung committed:
DOC: Add 0.23.2 whatsnew template (#21433)

gfyoung committed:
MAINT: More friendly error msg on Index overflow (#21377)

* MAINT: More useful error msg on Index overflow

Display a more friendly error message when there
is an OverflowError during Index construction.

Partially addresses gh-15832.

* DOC: Clarify how Index.__new__ handles dtype

Partially addresses gh-15823.

jorisvandenbossche committed:
DOC: follow 0.23.1 template for 0.23.2 whatsnew (#21435)

datapythonista committed:
DOC: Loading sphinxcontrib.spelling to sphinx only if it's available (#21397)

jorisvandenbossche committed:
Fix flake8 in conf.py (#21438)

alimcmaster1 committed:
Doc Fixes (#21415)

testvinder committed:
Reapply all patches by @testvinder against master (#21413)

akaihola committed:
Two tests didn't properly assert an exception was raised. Fixed. (#21409)

TomAugspurger committed:
DOC: 0.23.1 release (#21446)

TomAugspurger committed:
DOC: Fixed warning in doc build (#21449)

joelostblom committed:
DOC: add favicon to doc pages (#21440)

alimcmaster1 committed:
Fix tests fragile to PATH (#21453)

jbrockmendel committed:
use ccalendar.get_days_in_month, deprecate tslib.monthrange (#21451)

mroeschke committed:
BUG: Construct Timestamp with tz correctly near DST border (#21407)

jbrockmendel committed:
parametrize tests, unify repeated tests (#21405)

beepscore committed:
DOC: isin() docstring change DataFrame to pd.DataFrame (#21403)

toobaz committed:
BUG: fix get_indexer_non_unique with CategoricalIndex key (#21457)

closes #21448

jbrockmendel committed:
BUG: Fix DateOffset eq to depend on normalize attr (#21404)

Partially addresses gh-18854

mroeschke committed:
API/BUG: DatetimeIndex correctly localizes integer data  (#21216)

peterpanmj committed:
PERF: improve performance of groupby rank (#21237) (#21285)

jbrockmendel committed:
PERF: typing and cdefs for tslibs.resolution (#21452)

jbrockmendel committed:
disallow normalize=True with Tick classes (#21427)

KalyanGokhale committed:
CLN: Comparison methods for MultiIndex should have consistent behaviour for all nlevels  (GH21149) (#21195)

topper-123 committed:
PERF: Add __contains__ to CategoricalIndex (#21369)

mroeschke committed:
CLN: Index imports and 0.23.1 whatsnew (#21490)

topper-123 committed:
improve speed of nans in CategoricalIndex (#21493)

jbrockmendel committed:
perf improvements in tslibs.period (#21447)

jschendel committed:
BUG: Fix Series.nlargest for integer boundary values (#21432)

uds5501 committed:
Removing SimpleMock test from pandas.util.testing (#21482)

uds5501 committed:
TST: adding test cases for verifying correct values shown by pivot_table() #21378 (#21393)

alimcmaster1 committed:
Fix Timestamp rounding

pep8speaks commented:
Hello @alimcmaster1! Thanks for submitting the PR.

 - In the file [`pandas/tests/scalar/timestamp/test_unary_ops.py`](https://github.com/pandas-dev/pandas/blob/ee8ba611a8e3b0b081e7a74fe596ec8e0dbc16ab/pandas/tests/scalar/timestamp/test_unary_ops.py), following are the PEP8 issues :

> [Line 122:38](https://github.com/pandas-dev/pandas/blob/ee8ba611a8e3b0b081e7a74fe596ec8e0dbc16ab/pandas/tests/scalar/timestamp/test_unary_ops.py#L122): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 123:38](https://github.com/pandas-dev/pandas/blob/ee8ba611a8e3b0b081e7a74fe596ec8e0dbc16ab/pandas/tests/scalar/timestamp/test_unary_ops.py#L123): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 124:38](https://github.com/pandas-dev/pandas/blob/ee8ba611a8e3b0b081e7a74fe596ec8e0dbc16ab/pandas/tests/scalar/timestamp/test_unary_ops.py#L124): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 127:38](https://github.com/pandas-dev/pandas/blob/ee8ba611a8e3b0b081e7a74fe596ec8e0dbc16ab/pandas/tests/scalar/timestamp/test_unary_ops.py#L127): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 283:56](https://github.com/pandas-dev/pandas/blob/ee8ba611a8e3b0b081e7a74fe596ec8e0dbc16ab/pandas/tests/scalar/timestamp/test_unary_ops.py#L283): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

alimcmaster1 committed:
Pep8 fixes

mroeschke commented:
A few thoughts:

1) With your solution, I think it's better to make `round_ns` only take arrays (therefore wrap scalar inputs as array) to avoid checking the type of the input (and moreover `cdef`ing them) and iterating over values. We do something similar with `tz_localize_to_utc`.

2) Include tests for `DatetimeIndex.round`

3) Looks like there are some linting errors (your spacing seems to be 2 instead of 4 in some places).

alimcmaster1 committed:
Pep8 fixes and add additional test cases

alimcmaster1 committed:
Futher test cases

alimcmaster1 commented:
Thanks @mroeschke for the comments. I've done (2) and (3) added a few additional test cases to the DateTimeIndex ceil/floor and one that clearly shows how round should behave for this bug.

1) Yes this definitely makes sense. Just to clarify are you thinking in 
`_round` we check the type of `value` if its an int we wrap in an np array. Hence we can cdef `round_ns`?

Best,

Alistair

mroeschke commented:
Rather have `round_ns` (which we can at best `cpdef` since it's imported by a python file) _only_ accept numpy arrays. So since this method is used by timestamps, we would have to wrap it's `self.value` input as a numpy array. The relevant rounding code for `Timestamp`s would turn into:

```
def _round(self, freq, rounder):
    if self.tz is not None:
        value = self.tz_localize(None).value
    else:
        value = self.value
    value = np.array([value], dtype=np.int64)
    r = round_ns(value, rounder, freq)
```

mroeschke commented on pandas/tests/scalar/timestamp/test_unary_ops.py:
```diff
@@ -118,6 +118,27 @@ def test_ceil_floor_edge(self, test_input, rounder, freq, expected):
+        dt = Timestamp(test_input)
+        expected = Timestamp(expected)
+
+        result_ceil = dt.ceil(freq)
```
Could you also parametrize over the rounding methods (`ceil`, `floor`, and `round`)? It would help reduce this duplication

alimcmaster1 replied:
Sure makes sense to me

toobaz committed:
PERF: remove useless overrides (#21523)

closes #21522

mroeschke committed:
TST: Add unit tests for older timezone issues (#21491)

TomAugspurger committed:
BUG: Timedelta.__bool__ (#21485)

Closes #21484

Liam3851 committed:
BUG: Fix Index construction when given empty generator (#21470). (#21481)

minggli committed:
BUG/REG: file-handle object handled incorrectly in to_csv (#21478)

WillAyd committed:
Append Mode for ExcelWriter with openpyxl (#21251)

topper-123 committed:
DOC: Improve code example for Index.get_indexer (#21511)

adamjstewart committed:
DOC: remove grammar duplication in groupby docs (#21534)

jbrockmendel committed:
remove daytime attr, move getstate and setstate to base class (#21533)

r00ta committed:
BUG: Handle read_csv corner case (#21176)

Closes gh-21141

toobaz committed:
De-duplicate code for indexing with list-likes of keys (#21503)

wil committed:
Update "See Also" section of pandas/core/generic.py (#21550)

datapythonista committed:
Fixing documentation lists indentation (#21519)

topper-123 committed:
PERF: add method Categorical.__contains__ (#21508)

KalyanGokhale committed:
REGR: Fixes first_valid_index when DataFrame or Series has duplicate row index (GH21441) (#21497)

gfyoung committed:
API/BUG: Raise when int-dtype coercions fail (#21456)

Closes gh-15832.

Safrone commented:
So it shows up on the issue properly: #21262

jschendel committed:
DOC: Add documentation for freq='infer' option of DatetimeIndex and TimedeltaIndex constructors (#21566)

alimcmaster1 committed:
Refactor timestamp rounding

alimcmaster1 committed:
Parameterize test cases

alimcmaster1 commented:
Cleaned up my implementation here
1) Params added in test cases
2) round_ns now only takes np.array ( hence cleaned up the logic in here )

Thoguhts @mroeschke ?

alimcmaster1 committed:
Update function error

jreback commented on pandas/_libs/tslibs/timestamps.pyx:
```diff
@@ -59,43 +59,54 @@ cdef inline object create_timestamp_from_ts(int64_t value,
-    return r
+        return r
+
+    return np.fromiter((_round_non_int_multiple(item) for item in values), np.int64)
```
why are you iterating? this doesn’t make any sense to do so with. vectorizes function

Commit:
BUG: Fix group index calculation to prevent hitting maximum recursion depth (#21541)

mroeschke commented on pandas/_libs/tslibs/timestamps.pyx:
```diff
@@ -59,43 +59,54 @@ cdef inline object create_timestamp_from_ts(int64_t value,
     ----------
-    values : int, :obj:`ndarray`
-    rounder : function
+    values : np.array
```
I'd leave this as :obj:`ndarray`

mroeschke commented on pandas/_libs/tslibs/timestamps.pyx:
```diff
@@ -59,43 +59,54 @@ cdef inline object create_timestamp_from_ts(int64_t value,
     Returns
     -------
-    int or :obj:`ndarray`
+    np.array
```
Same

mroeschke commented on pandas/_libs/tslibs/timestamps.pyx:
```diff
@@ -649,7 +660,10 @@ class Timestamp(_Timestamp):
+        value = np.array([value], dtype=np.int64)
+
+        # Will only ever contain 1 element for timestamp
+        r = round_ns(value, rounder, freq).item()
```
Nit. I think just indexing into this array (i.e. `round_ns(value, rounder, freq)[0]`) is just fine. Looks like `item` returns a copy of a Python scalar (and we may want to keep this a numpy scalar just in case)

mroeschke replied:
`self.value` from `Timestamp` will always be a scalar, so we implicitly know the result of this is be a one element array.

mroeschke commented:
Thanks for the revision. As @jreback mentions, I dont think it's necessary to iterate, You should be able to perform the adjustments with vectorization. At a high level the logic should look like:

```
mask = value % unit == 0
if mask.all():
    return value
values[~mask] = _round_non_int_multiple(values[~mask])
return values
```

alimcmaster1 committed:
BUG: Fix passing empty label to df drop (#21515)

Closes #21494

toobaz committed:
ERR: Raise a simpler backtrace for missing key (#21558)

spott committed:
Fixed HDFSTore.groups() performance.   (#21543)

datapythonista committed:
DOC: Fixing spaces around backticks, and linting (#21570)

jbrockmendel committed:
fix hashing string-casting error (#21187)

jbrockmendel committed:
make DateOffset immutable (#21341)

toobaz committed:
REF: multi_take is now able to tackle all list-like (non-bool) cases (#21569)

alimcmaster1 commented:
That @mroeschke your logic above seems much neater let me refactor!

gfyoung committed:
TST: Use int fixtures in test_construction.py (#21588)

Partially addresses gh-21500.

uds5501 committed:
DOC: Adding clarification on return dtype of to_numeric (#21585)

uds5501 committed:
Update v0.24.0.txt (#21586)

mitchnegus committed:
clarifying regex pipe behavior (#21589)

gfyoung committed:
DOC: Note assert_almost_equal impl. detail (#21580)

Note the hard-coded switch between absolute
and relative tolerance during checking.

Closes gh-21528.

strickvl committed:
DOC: update the Series.any / Dataframe.any docstring (#21579)

gfyoung committed:
TST: Clean up tests in test_take.py (#21591)

Utilizes pytest's fixtures to
cleanup the code significantly.

jschendel committed:
TST: Add interval closed fixture to top-level conftest (#21595)

jbrockmendel committed:
cache DateOffset attrs now that they are immutable (#21582)

minggli committed:
BUG: Series dot product __rmatmul__ doesn't allow matrix vector multiplication (#21578)

reidy-p committed:
BUG: first/last lose timezone in groupby with as_index=False (#21573)

r00ta committed:
add test case when to_csv argument is sys.stdout (#21572)

vuminhle committed:
BUG: Fix json_normalize throwing TypeError (#21536) (#21540)

ryankarlos committed:
DOC: updated the Series.str.rsplit and Series.str.split docstrings (#21026)

jschendel committed:
TST: Use multiple instances of parametrize instead of product (#21602)

WillAyd committed:
MyPy cleanup and absolute imports in pandas.core.dtypes.common (#21008)

alimcmaster1 committed:
Perform manipulation with vectorization

mroeschke commented on pandas/_libs/tslibs/timestamps.pyx:
```diff
@@ -59,43 +59,58 @@ cdef inline object create_timestamp_from_ts(int64_t value,
-    values : int, :obj:`ndarray`
-    rounder : function
+    values : :obj:`ndarray`
+    rounder : function, eg. 'Ceil', 'Floor', 'round'
```
Nit: Could you lowercase ceil and floor

mroeschke commented on pandas/_libs/tslibs/timestamps.pyx:
```diff
@@ -59,43 +59,58 @@ cdef inline object create_timestamp_from_ts(int64_t value,
 
-        r = (unit * rounder((values * (divisor / float(unit))) / divisor)
-             .astype('i8'))
+    values = np.copy(values)
```
Why do we need a copy here?

alimcmaster1 replied:
This is to handle the case where 'NaT' exists in the DateTimeIndex. Removing it will cause `test_ceil_floor_edge` in `test_scalar_compact.py` to fail.

I found that datetimelike.py `self.hasnans` in `_maybe_mask_results`will return False if we don't do the copy, by copying we ensure that we arn't referencing the base of the input array. Think that is the issue here, what do you think?

mroeschke replied:
cc @jreback.

Thanks for the investigation. That sounds reasonable; but I am not too familiar with `nan`/`Nat` ops with respect to references.

jreback replied:
you are mutating in place, so you DO need to copy.

that's fine. though pls use ``values.copy()``

jbrockmendel committed:
remove unused cimport (#21619)

TomAugspurger committed:
CI: Test against Python 3.7 (#21604)

alimcmaster1 committed:
Lower case doc string

aberres committed:
DOC: Do no use 'type' as first word when specifying a return type (#21622) (#21623)

topper-123 committed:
CLN: make CategoricalIndex._create_categorical a classmethod (#21618)

toobaz committed:
PERF: do not check for label presence preventively (#21594)

closes #21593

minggli committed:
TST: Refactor test_maybe_match_name and test_hash_pandas_object (#21600)

jreback reviewed (changes requested):
i haven't fully reviewd yet

TomAugspurger committed:
API/COMPAT: support axis=None for logical reduction (reduce over all axes) (#21486)

* Compat with NumPy 1.15 logical func
* Accepts axis=None as reduce all dims

mroeschke committed:
DOC: Move tz cleanup whatsnew entries to v0.24 (#21631)

jorisvandenbossche committed:
DOC: fixup old whatsnew for dtype coercing change (#21456) (#21634)

KalyanGokhale committed:
DEPR: MultiIndex.to_hierarchical (#21613)

jreback committed:
TST: xfail flaky 3.7 test, xref #21636 (#21637)

bnaul committed:
[ENH] Add read support for Google Cloud Storage (#20729)

* Google Cloud Storage support using gcsfs

TomAugspurger committed:
PKG: Exclude data test files. (#19535)

bgroveben committed:
DOC: fix typo in cookbook.rst (#21635)

Removing the semicolon delimiter at the end of the modified line of code allows the line's output to be displayed.

alimcmaster1 committed:
Update copy function

alimcmaster1 commented:
Thanks @jreback, ive updated, let me know if there is anything else!

jreback commented on pandas/_libs/tslibs/timestamps.pyx:
```diff
@@ -59,43 +59,58 @@ cdef inline object create_timestamp_from_ts(int64_t value,
+    values = values.copy()
 
-    return r
+    # GH21262 If the Timestamp is multiple of the freq str
```
i think it simpler to put this condition inside ththe above function

then you don’t even need a function !

iow i think you just needed to add that case above  (in the original code)

alimcmaster1 replied:
Right yep I see what you are saying, just to confirm before I commit, something like [this](https://gist.github.com/alimcmaster1/824ea777641c5c54f58e4f279a334210) ?

jreback replied:
yep!

topper-123 committed:
DOC: minor correction to v0.23.2.txt  (#21644)

david-liu-brattle-1 committed:
Cleanup clipboard tests (#21163)

tswast committed:
ENH: Update to_gbq and read_gbq to pandas-gbq 0.5.0 (#21628)

* Add link to Pandas-GBQ 0.5.0 in what's new.
* Remove unnecessary sleep in GBQ tests.

Closes https://github.com/pydata/pandas-gbq/issues/177

Closes #21627

jbrockmendel committed:
More speedups for Period comparisons (#21606)

jbrockmendel committed:
use ccalendar instead of np_datetime (#21549)

alimcmaster1 committed:
Refactor to use just one function

jschendel committed:
DOC: Fix versionadded directive typos in IntervalIndex (#21649)

elmq0022 committed:
TST: Use absolute path for datapath (#21647)

taljaards committed:
DOC: update DataFrame.dropna's axis argument docs (#21652)

jschendel committed:
BUG: Let IntervalIndex constructor override inferred closed (#21584)

alimcmaster1 commented:
Done! @jreback

mroeschke commented:
Travis flagged some linting errors:

```
pandas/_libs/tslibs/timestamps.pyx:92:14: E128 continuation line under-indented for visual indent
pandas/_libs/tslibs/timestamps.pyx:92:80: E501 line too long (80 > 79 characters)
pandas/_libs/tslibs/timestamps.pyx:103:80: E501 line too long (87 > 79 characters)
pandas/_libs/tslibs/timestamps.pyx:104:14: E128 continuation line under-indented for visual indent
```

alimcmaster1 committed:
Pep8

gfyoung committed:
TST: Use fixtures in dtypes/test_cast.py (#21661)

mroeschke committed:
TST: Clean old timezone issues PT2 (#21612)

jreback reviewed (changes requested):
ok lgtm. can you add a whatsnew note in Fixed Regression for 0.23.2, ping on green.

alimcmaster1 committed:
Whatsnew Timestamp bug

alimcmaster1 committed:
Merge branch 'master' into timestamp-fixes

alimcmaster1 commented:
thanks @jreback and @mroeschke for helping review!
